### PR TITLE
builder API integration (for latest spec)

### PR DESCRIPTION
**Pull request by metachris**
* Updates relay implementation to conform with [mev-boost spec v0.2.1](https://github.com/flashbots/mev-boost/blob/main/docs/specification.md)
* Supersedes #20

### CODE REVIEW

File: consensus.go [hunk 1]
```diff
@@ -380,29 +383,23 @@ func (c *ConsensusCmd) RunNode() {
...
 }
 
-func (c *ConsensusCmd) sendForkchoiceUpdated(latest, safe, final common.Hash, attributes *PayloadAttributesV1) *PayloadID {
+func (c *ConsensusCmd) sendForkchoiceUpdated(latest, safe, final common.Hash, attributes *PayloadAttributesV1) (*PayloadID, error) {
 	result, _ := ForkchoiceUpdatedV1(c.ctx, c.engine, c.log, latest, safe, final, attributes)
 	if result.PayloadStatus.Status != ExecutionValid {
 		c.log.WithField("status", result.PayloadStatus).Error("Update not considered valid")
+		return nil, fmt.Errorf("update not considered valid")
 	}
-	if c.builder != nil && attributes != nil {
```
Comment (metachris) [inline consensus.go:388]: Removed because the builder API no longer supports `forkChoiceUpdated`

File: api/builder.go [hunk 2]
```diff
@@ -6,103 +6,51 @@ import (
 	"mergemock/rpc"
 
 	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/common/hexutil"
-	"github.com/ethereum/go-ethereum/core/types"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
 	gethRpc "github.com/ethereum/go-ethereum/rpc"
 	"github.com/ethereum/go-ethereum/trie"
-	"github.com/holiman/uint256"
+	"github.com/flashbots/mev-boost/types"
```
Comment (lightclient) [inline api/builder.go:12]: Please remove this dependency. We should redefine all the types in `mergemock`. This let's us release new versions without waiting on upstream components.

File: api/builder.go [hunk 3]
```diff
@@ -6,103 +6,51 @@ import (
...
 	"github.com/sirupsen/logrus"
 )
 
-//go:generate go run github.com/fjl/gencodec -type ExecutionPayloadHeaderV1 -field-override executionPayloadHeaderMarshalling -out gen_edh.go
+func BuilderGetHeader(ctx context.Context, cl *rpc.Client, log logrus.Ext1FieldLogger, blockHash common.Hash) (*types.ExecutionPayloadHeaderV1, error) {
```
Comment (lightclient) [inline api/builder.go:16]: This should be extended to take `slot` and `pubkey`.

File: api/builder.go [hunk 4]
```diff
@@ -6,103 +6,51 @@ import (
...
+				e.Warn("unavailable payload in get-payload header request")
 			}
 		} else {
 			e.Error("failed to get payload header")
 		}
 		return nil, err
 	}
 	e.Debug("Received payload")
-	return &result, nil
+	return &result.Message.Header, nil
```
Comment (lightclient) [inline api/builder.go:38]: Should return either `message` and `signature` or it should verify `message` against `signature` and return `message` if successful.

File: api/builder.go [hunk 5]
```diff
@@ -6,103 +6,51 @@ import (
...
 }
 
-func ProposePayload(ctx context.Context, cl *rpc.Client, log logrus.Ext1FieldLogger, header *ExecutionPayloadHeaderV1) (*ExecutionPayloadV1, error) {
+func BuilderGetPayload(ctx context.Context, cl *rpc.Client, log logrus.Ext1FieldLogger, header *types.ExecutionPayloadHeaderV1) (*ExecutionPayloadV1, error) {
```
Comment (lightclient) [inline api/builder.go:41]: This takes a `BlindedBeaconBlock` and a `signature`.

File: relay.go [hunk 6]
```diff
@@ -109,41 +105,51 @@ func (r *RelayCmd) startRPC(ctx context.Context, backend *RelayBackend) {
...
 	if err != nil {
 		return nil, err
 	}
-	return &RelayBackend{log, engine, 0, cache}, nil
+	return &RelayBackend{log, engine, cache}, nil
+}
+
+func (r *RelayBackend) RegisterValidatorV1(ctx context.Context, message types.RegisterValidatorRequestMessage, signature hexutil.Bytes) (*string, error) {
+	res := "OK"
+	return &res, nil
```
Comment (lightclient) [inline relay.go:129]: This functions needs to correctly register the validator, including verifying the signature.

File: relay.go [hunk 7]
```diff
@@ -109,41 +105,51 @@ func (r *RelayCmd) startRPC(ctx context.Context, backend *RelayBackend) {
...
-	return payload.(*ExecutionPayloadHeaderV1), nil
+	return &types.GetHeaderResponse{Message: types.GetHeaderResponseMessage{Header: *payloadHeader, Value: (*hexutil.Big)(val)}}, nil
 }
 
-func (r *RelayBackend) ProposeBlindedBlockV1(ctx context.Context, block *SignedBlindedBeaconBlock, attributes *SignedBuilderReceipt) (*ExecutionPayloadV1, error) {
-	// TODO: The signed messages should be verified. It should ensure that the signed beacon block is for a validator
-	// in the expected slot. The attributes should be verified against the relayer's key.
-	hash := block.Message.Body.ExecutionPayload.BlockHash
-	plog := r.log.WithField("payload_hash", hash)
+func (r *RelayBackend) GetPayloadV1(ctx context.Context, block types.BlindedBeaconBlockV1, signature hexutil.Bytes) (*ExecutionPayloadV1, error) {
```
Comment (lightclient) [inline relay.go:150]: Should only reveal if the block is correctly signed.

**Review by lightclient — CHANGES_REQUESTED** [review]
Missing SSZ of the types and BLS support.

**Pull request closed without merging**